Display help output on stdout #8980

miminar opened the pull request:
This is a small adjustment of the work done by @rhatdan and @tiborvass. It
suppresses redundant error messages and handles a case where docker is run with
a bad option without a command.

Fixes #8980 and closes #7887

rhatdan committed:
Remove TestUsage, since Usage will no longer be shown on failure to parse

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

rhatdan committed:
--help option and help command should print to stdout not stderr

    --help and help are successful commands so output should not go to error.

    QE teams have requested this change, also users doing docker help | less
    or docker run --help | less would expect this to work.

    Usage statement should only be printed when the user asks for it.
    Errors should print error message and then suggest the docker COMMAND --help
    command to see usage information.

    The current behaviour causes the user to have to search for the error message
    and sometimes scrolls right off the screen.  For example a error on a
    "docker run" command is very difficult to diagnose.

    Finally erros should always exit with a non 0 exit code, if the user
    makes a CLI error.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

tiborvass committed:
refactor redundant code around calls to cmd.Parse

Signed-off-by: Tibor Vass <teabee89@gmail.com>

Commit:
Handle bad options better

* Do not log bad options error message twice, e.g.:

    $ docker run --pouet
    flag provided but not defined: --pouet
    See 'docker run --help'.
    2014/11/05 21:41:23 flag provided but not defined: --pouet

  With this patch just the first two lines will be produced.

* Print 'docker' just once when run without a command, e.g.:

    $ docker --hel
    flag provided but not defined: --hel
    See 'docker docker --help'.

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Fixed error reporting

Removed redundant print line and fixed handling of command-less docker
invocation.

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Exit with non-zero code on first argument parsing error

Ignoring return value of ParseFlags leads to exit code 0 if bad
arguments are supplied. This patch makes sure that subcommands exit
with non-zero code in such a case.

Signed-off-by: Michal Minar <miminar@redhat.com>

rhatdan commented on pkg/mflag/flag.go:
```diff
@@ -788,8 +851,12 @@ func Var(value Value, names []string, usage string) {
+		fmt.Fprintf(f.Out(), "See '%s --help'.\n", os.Args[0])
+	} else {
+		fmt.Fprintf(f.Out(), "See '%s %s --help'.\n", os.Args[0], f.name)
+	}
```
I totally disagree.

We should follow the coreutils standard

ls --fobar
ls: unrecognized option '--fobar'
Try 'ls --help' for more information.

Only display usage if user asks for it.

Commit:
Amended TestExecParseError test

Usage string isn't printed anymore. User is adviced to see help instead
- according to coreutils standard.

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
@jfrazelle: Rebase done. For the last time I hope...
Thanks for taking care of it.

rhatdan commented:
We have only been working on this one for 9 months now...

jfrazelle commented:
great LGTM, ping @tiborvass :)

tiborvass commented:
LGTM

If we discover other inconsistencies, we can fix them in a separate PR. I agree this took way longer than all of us wanted. Thanks everyone for your patience.
